src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorResolver.cs

```diff
@@ -124,6 +133,79 @@ public IEnumerable<TagHelperDescriptor> Resolve([NotNull] TagHelperDescriptorRes
             return descriptors;
         }
 
+        private static IEnumerable<TagHelperDescriptor> PrefixDescriptors(
+            TagHelperDescriptorResolutionContext context,
+            IEnumerable<TagHelperDescriptor> descriptors)
+        {
+            var tagHelperPrefix = ResolveTagHelperPrefix(context);
+
+            if (!string.IsNullOrEmpty(tagHelperPrefix))
+            {
+                return descriptors.Select(descriptor =>
+                    new TagHelperDescriptor(
+                        tagHelperPrefix,
+                        descriptor.TagName,
+                        descriptor.TypeName,
+                        descriptor.AssemblyName,
+                        descriptor.Attributes));
+            }
+
+            return descriptors;
+        }
+
+        private static string ResolveTagHelperPrefix(TagHelperDescriptorResolutionContext context)
+        {
+            var prefixDirectiveDescriptors = context.DirectiveDescriptors.Where(
+                descriptor => descriptor.DirectiveType == TagHelperDirectiveType.TagHelperPrefix);
+
+            // For each invalid @tagHelperPrefix we need to create an error.
+            foreach (var erroredContext in prefixDirectiveDescriptors.Skip(1))
+            {
+                context.ErrorSink.OnError(
+                    erroredContext.Location,
+                    Resources.FormatTagHelperDescriptorResolver_InvalidTagHelperDirective(
+                        SyntaxConstants.CSharp.TagHelperPrefixKeyword));
+            }
+
+            var prefixDirective = prefixDirectiveDescriptors.FirstOrDefault();
+            var prefix = prefixDirective?.DirectiveText;
+
+            if (prefix == null || !EnsureValidPrefix(prefix, prefixDirective.Location, context.ErrorSink))
+            {
+                prefix = string.Empty;
+            }
+
+            return prefix;
+        }
+
+        private static bool EnsureValidPrefix(
+            string prefix,
+            SourceLocation directiveLocation,
+            ParserErrorSink errorSink)
+        {
+            foreach (var character in prefix)
+            {
+                // Prefixes are correlated with tag names, tag names cannot have whitespace.
+                if (char.IsWhiteSpace(character) ||
+                    // Reserved Razor transition symbol
+                    character == '@' ||
+                    // Reserved TagHelper opt-out character.
+                    character == '!')
```

Comment:
from `HtmlSymbolType` I'd expect this blacklist (generally poor idea despite the existing mess w.r.t. character exclusion but :ok: for now) to be much longer.  should at least exclude everything that tokenizer will ensure is not found in a `HtmlSymbolType.Text` token.